MetaLamp/lending-pool/src/Plutus/State/Update.hs

```diff
@@ -0,0 +1,122 @@
+import qualified Data.Text                        as Text
+import           Data.Void                        (Void)
+import           Ledger                           hiding (getDatum, singleton)
+import           Ledger.Constraints               as Constraints
+import           Ledger.Constraints.OnChain       as Constraints
+import           Ledger.Constraints.TxConstraints as Constraints
+import qualified Ledger.Scripts                   as Scripts
+import           Ledger.Typed.Scripts             (ScriptType (..))
+import qualified Ledger.Typed.Scripts             as Scripts
+
+import           Playground.Contract
+import           Plutus.Contract                  hiding (when)
+import           Plutus.State.Select              (StateOutput (..))
+import           Plutus.V1.Ledger.Value
+import           PlutusTx                         (IsData)
+import qualified PlutusTx
+import           PlutusTx.Prelude                 hiding (Semigroup (..),
+                                                   unless)
+import           Prelude                          (Semigroup (..))
+import qualified Prelude
+
+type RootToken = AssetClass
+
+-- State token can be only be forged either when input has a root token(first-time creation)
+-- or any of the state tokens(modification), assuming that state token was created with a root token at some point
+{-# INLINABLE validateStateForging #-}
+validateStateForging :: RootToken -> TokenName -> ScriptContext -> Bool
+validateStateForging rootToken tn ctx = case [ i
+                                          | i <- txInfoInputs $ scriptContextTxInfo ctx
+                                          , let v = valueWithin i
+                                          , (assetClassValueOf v rootToken == 1) ||
+                                            (assetClassValueOf v stateToken == 1)
+                                          ] of
+    [_]    -> True
+    [_, _] -> True
+    _      -> traceError "State forging without RootToken input"
+  where
+    stateToken = assetClass (ownCurrencySymbol ctx) tn
+    valueWithin = txOutValue . txInInfoResolved
+
+makeStatePolicy :: RootToken -> TokenName -> MonetaryPolicy
+makeStatePolicy rootToken tokenName = mkMonetaryPolicyScript $
+    $$(PlutusTx.compile [|| \r t -> Scripts.wrapMonetaryPolicy (validateStateForging r t) ||])
+        `PlutusTx.applyCode` PlutusTx.liftCode rootToken
+        `PlutusTx.applyCode` PlutusTx.liftCode tokenName
+
+makeStateCurrency :: RootToken -> TokenName -> CurrencySymbol
+makeStateCurrency rootToken = scriptCurrencySymbol . makeStatePolicy rootToken
+
+makeStateToken :: RootToken -> TokenName -> AssetClass
+makeStateToken rootToken tokenName = assetClass (makeStateCurrency rootToken tokenName) tokenName
+
+data RootStateHandle a b = RootStateHandle {
+    script    :: Scripts.ScriptInstance a,
+    rootToken :: AssetClass,
+    output    :: StateOutput b
+}
+
+data StateHandle i o a = StateHandle {
```

Comment:
could we parametrize `data StateHandle i o a ` by fewer type vars, it seems like we could reduce two of them 
`i = RedeemerType t ` `o = DatumType t`